File: Algorithms/139.WordBreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string> &wordDict) {
        int n = s.size();
        bool wb[n];
        memset(wb, false, n);
        for (int i = 0; i < n; ++i) {
            if (wb[i] == false && find(begin(wordDict), end(wordDict), s.substr(0, i+1)) != wordDict.end())
                wb[i] = true;
            if (wb[i] == true) {
                if (i == n-1) return true;
                for (int j = i+1; j < n; ++j) {
                    if (wb[j] == false && 
                            find(begin(wordDict), end(wordDict), s.substr(i+1, (j+1)-(i+1))) != wordDict.end())
                        wb[j] = true;
                    if (j == n-1 && wb[j] == true) return true;
                }
            }
        }
        return false;
    }
};

File: Algorithms/py/139.WordBreak.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        n = len(s)
        wb = [False] * n
        for i in range(n):
            if wb[i] == False and s[:i+1] in wordDict:
                wb[i] = True
            if wb[i]:
                if i == n-1: return True
                for j in range(i+1, n):
                    if wb[j] == False and s[i+1:j+1] in wordDict:
                        wb[j] = True
                    if j == n-1 and wb[j]:return True
        return False
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n+1)
        dp[0] = True
        for i in range(n+1):
            for j in range(i, n+1):
                if dp[i] and s[i:j] in wordDict:
                    dp[j] = True
        return dp[-1]